### ISSUE TITLE
BUG: Simultaneous access to different DataFrame instances by different threads should be safe

### ISSUE THREAD

**Issue by ptomask** [pydata/pandas#12420]
I need to perform computations in Pandas in multiple threads. None of the dataframes are shared across running threads. Although there is some race condition witch causes the interpreter to crash.

I Googled for help and [found this](http://stackoverflow.com/questions/25782912/pandas-and-numpy-thread-safety).

The snippet included in that stackoverflow discussion is definitely crashing my interpreter, so I insert it below as a bug reproduction.

I am aware of Pandas not being thread-safe, but thread-safety feature mostly considers the case of accessing the same object instance by more threads. Threads in the snippet below, however, have all their own DataFrame instance, so the thread safety at least theoretically should not be needed.

#### Code Sample, a copy-pastable example if possible
```
import threading
import pandas as pd
import numpy as np

def let_crash(crash=True):
    t = 0.02 * np.arange(100000) # ok con 10000                                                                               
    data = pd.DataFrame({'t': t})
    if crash:
        data['t'] * 1.5  # CRASH
    else:
        data['t'].values * 1.5  # THIS IS OK!

if __name__ == '__main__':
        threads = []
        for i in range(100):
            if True:  # asynchronous                                                                                          
                t = threading.Thread(target=let_crash, args = ())
                t.daemon = True
                t.start()
                threads.append(t)
            else:  # synchronous                                                                                              
                let_crash()
        for t in threads:
            t.join()
```

#### Expected Output
The code should just finish without the interpreter dying by access violation error.

#### output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 32
OS: Windows
OS-release: 8
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.16.2
nose: 1.3.7
Cython: 0.22.1
numpy: 1.9.2
scipy: 0.15.1
statsmodels: 0.6.1
IPython: 2.4.1
sphinx: 1.3.2
patsy: 0.3.0
dateutil: 2.4.2
pytz: 2015.4
bottleneck: 1.0.0
tables: 3.2.0
numexpr: 2.4.3
matplotlib: 1.4.3
openpyxl: 2.2.4
xlrd: 0.9.3
xlwt: 1.0.0
xlsxwriter: None
lxml: 3.4.4
bs4: 4.3.2
html5lib: 0.99999
httplib2: None
apiclient: None
sqlalchemy: 1.0.6
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext)

#### Windows error message
The error message changes occassionally, I included two versions below. Also I am translating from Czech Windows version so the original English field names probably vary.

```
Problem description:
  Problem event name:	APPCRASH
  Application name:	python.exe
  Application version:	0.0.0.0
  Application timestamp:	5560ad83
  Name of errorneous module:	python27.dll
  Version of errorneous module:	2.7.10150.1013
  Timestamp of errorneous module:	5560ad81
  Exception code:	c0000005
  Exception shift:	000ae4a9
  Operating system version:	6.3.9600.2.0.0.768.101
  National environment ID:	1029
  Further info 1:	5861
  Further info 2:	5861822e1919d7c014bbb064c64908b2
  Further info 3:	1318
  Further into 4:	13181ae637ac2592fac0ca2dd662cdcc
```

```
Problem description:
  Problem event name:	APPCRASH
  Application name:	python.exe
  Application version:	0.0.0.0
  Application timestamp:	5560ad83
  Name of errorneous module:	interpreter.pyd
  Version of errorneous module:	0.0.0.0
  Timestamp of errorneous module:	5543308a
  Exception code:	c0000005
  Exception shift:	00007e6c
  Operating system version:	6.3.9600.2.0.0.768.101
  National environment ID:	1029
  Further info 1:	5861
  Further info 2:	5861822e1919d7c014bbb064c64908b2
  Further info 3:	1318
  Further info 4:	13181ae637ac2592fac0ca2dd662cdcc
```

**Comment by jreback** [comment]
so, I believe all of the issues w.r.t. threading are due to ``numexpr<2.4.6`` (there is another fix in 2.5 so might has well run on that. Further you have an older version of pandas. v0.18.0 coming shortly fixes even more issues. you can install via: ``conda install pandas=v0.18.0rc1 -c pandas``

I confirm this works on master on windows 3.5 with numexpr=2.5.

**Comment by jreback** [comment]
as an aside. no reason to run 32-bit windows nowadays. you can always install your own conda environment.

**Comment by ptomask** [comment]
Thank you for your ideas. I performed all the upgrades you recommended. Unfortunatelly the problem just moved somewhere else. The snippet above already works correctly, but my code still crashes the interpreter. I spared quite some hours trying to create a new snippet to reproduce the bug, but I can still observe the behavior only in original codebase of my project. The bug is very sensitive to timing, e.g. when I add some printing to the suspicious parts of the code, the error is gone...
I will post the new snippet as soon as I am able to create one.
